docs/pipeline-yml-style.md

```diff
@@ -0,0 +1,112 @@
+
+> ![](images/one-stage-per-job.png)
+
+> Related feedback item: [Allow failed Jobs to be retried as soon as they have finished running - Visual Studio Feedback](https://developercommunity.visualstudio.com/t/Allow-failed-Jobs-to-be-retried-as-soon/10130213)
+
+> Also requested internally (2018): https://dev.azure.com/dnceng/internal/_queries/edit/110
+
+## Dynamic vs. typed template parameter declarations
+
+Most of our templates use dynamic template parameters. These look like:
+
+```yml
+parameters:
+  apple: 2
+  lemon: "acidic"
+  orange: true
+```
+
+In this mode, additional parameters can be passed when using a `- template: ...` statement. All parameters other than objects and lists seem to be typecast to `string` when passed in. This means `${{ if parameters.orange }}` will pass if `orange` is passed in as `false`, because the implicitly converted `"False"` is truthy.
+
+A workaround is to instead use `${{ if eq(parameters.orange, true) }}` to (it seems) perform the implicit cast on both sides.
+
+Some templates instead use typed parameters:
+
+```yml
+parameters:
+  - name: orange
+    type: boolean
+    default: true
+```
+
+In this mode, any additional parameters passed by a `- template: ...` statement are rejected as a template evaluation error. If we need some arbitrary parameters, we use an extra `object` param.
+
+Because `orange` is passed as a `boolean`, `${{ if parameters.orange }}` evaluates as expected.
+
+## Templates for data reuse
+
+AzDO supports [variable templates](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops#variable-templates-with-parameter), but it's hard to determine where variables defined this way are usable. For example, a macro expansion `$(example)` will evaluate to define a job's pool name, but not the demands on that pool. A runtime expression `$[example]` will not evaluate in either case. A template expression `${{ example }}` will always work.
+
+For universal applicability, we use templates to share logic and values between stages. Another benefit is early evaluation: we can sometimes catch errors in the queue dialog (rather than at build-time), saving dev time.
+
+## Editing tools
+
+The [Azure Pipelines extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-azure-devops.azure-pipelines) fails to parse our templates properly. It doesn't generally understand template expressions, doesn't understand dynamic template parameters, and the language server crashes when it sees `insert` as the first element in a list, like this:
+
+```yml
+- ${{ each value in parameters.thing }}:
+  - ${{ insert }}: ${{ value }}
+    answer: 42
+```
+
+We edit the YAML files with generic YAML tools.
+
+## Indentation style
+
+The YAML files in this repository indent when starting a list, which isn't typical for pipeline YAML. This isn't specific to AzDO, but 
```

Comment:
Just noticed this--it kinda works as a lead-in for the example, but that wasn't what I intended to have here. 😄 

```suggestion
We indent when starting a new list in the YAML files in this repository. This isn't typical for pipeline YAML: in AzDO docs, a list's `-` is on the same column as the element containing the list. Both styles are commonly found in general YAML files. For example, we use:
```